design/one-pager-crossplane-state-metrics.md

```diff
@@ -0,0 +1,640 @@
+
+**Why this approach:**
+
+Different organizations have vastly different requirements:
+- A small team might only need basic resource counts
+- An enterprise might need team attribution across 50+ teams
+- A service provider might need customer-level isolation
+- A platform team might need SLOs per application team
+
+An external operator allows each organization to choose their trade-offs without impacting others.
+
+## Proposal
+
+We propose implementing a **State Metrics Operator for Crossplane**, a standalone controller that observes resources and exposes their state as configurable metrics.
+
+This operator will act as the metrics translation layer between Crossplane and external monitoring systems. By moving advanced metrics handling out of Crossplane Core and into an external, configurable operator, we preserve Crossplane's simplicity while unlocking flexibility for organizations with diverse observability requirements.
+
+### Architecture
+
+The operator is deployed as an independent controller with its own RBAC, CRDs, and reconciliation loops.
+
+**Core Components:**
+
+1. **ClusterMetric Controller**
+   - Watches ClusterMetric CRDs for configuration changes
+   - Discovers and monitors cluster-scoped target resources (Providers, cluster XRs, etc.)
+   - Extracts values and labels using JSONPath expressions
+   - Registers Prometheus metrics based on ClusterMetric definitions
+   - Tracks state transitions for counter metrics
+
+2. **Metric Controller**
+   - Watches Metric CRDs (namespace-scoped) for configuration changes
+   - Discovers and monitors namespace-scoped target resources (XRs, Claims, etc.)
+   - Same extraction and metric generation logic as ClusterMetric
+   - Provides namespace-level isolation for metrics
+
+3. **Prometheus Metrics Endpoint**
+   - Exposes metrics at `/metrics` endpoint
+   - Standard Prometheus scrape target
+   - Serves both gauge (current state) and counter (state transitions) metrics
+   - All metrics follow pattern: `crossplane_{metadata.name}` and `crossplane_{metadata.name}_total`
+
+### API Design
+
+#### Metric and ClusterMetric
+
+The operator provides two CRD types for defining metrics:
+
+- **Metric**: Namespace-scoped, for monitoring resources within a specific namespace
+- **ClusterMetric**: Cluster-scoped, for monitoring cluster-wide resources
+
+Both share the same core spec structure:
+
+```yaml
+apiVersion: metrics.crossplane.io/v1alpha1
+kind: ClusterMetric
+metadata:
+  name: xnetwork-ready-status
+spec:
+  # Help text for the metric (Prometheus HELP)
+  help: "XNetwork ready status (1=ready, 0=not ready)"
+
+  # Reconciliation interval (default: 10m)
+  interval: 1m
+
+  # Target resource to monitor
+  target:
+    group: aws.platform.upbound.io
+    kind: XNetwork        # Supports wildcard "*" for all kinds
+    version: v1alpha1
+
+  # Optional: Filter resources
+  labelSelector: "team=platform"
+  fieldSelector: "metadata.namespace=production"
```

Comment:
Nit: should this be a map, for better readability and to allow multiple selectors?